Make app F-Droid ready

Ch4t4r opened the pull request:
This PR aims to remove closed-source dependencies (or pre-compiled binaries from non-approved sources) to make this app ready for F-Droid.
I like to share my progress early as a PR to show progress and get feedback.

Right now this PR removes the play-services-core library from the assembled app and only includes it when an .apk file compiled from `./gradlew assemblePlayRelease` is used. For other, non-play builds it uses the server for checking for updates. This way folks at F-Droid could compile without the proprietary library bundled. 

Things left to do:
 - ~~Compile & test app~~
    - ~~Verify both updaters are called properly~~
 - ~~Extract other dependencies~~
 - ~~fix copyright header~~
 - ~~refactor~~

ignoramous commented on app/src/main/java/com/celzero/bravedns/NonStoreAppUpdater.kt:
```diff
@@ -0,0 +1,83 @@
+                    val responseVersion = jsonObject.getInt("version")
+                    val updateValue = jsonObject.getBoolean("update")
+                    val latestVersion = jsonObject.getInt("latest")
+                    persistentState.lastAppUpdateCheck = System.currentTimeMillis()
```
Please add a `FIXME` to move this to fetch network-time (NTP) where available.

ignoramous commented on app/src/main/java/com/celzero/bravedns/NonStoreAppUpdater.kt:
```diff
@@ -0,0 +1,83 @@
+        })
+    }
+
+    override fun completeUpdate() {
+    }
+
+    override fun unregisterListener(listener: AppUpdater.InstallStateListener) {
+    }
```
nit: please consider adding no-op comment (since it was intentionally left blank)

ex:
```
override fun forcedImpl() {
    // no-op
}

```

ignoramous reviewed (commented):
Thanks! :D 

cc: @hussainmohd-a

Ch4t4r commented on app/src/main/java/com/celzero/bravedns/ui/HomeScreenActivity.kt:
```diff
@@ -73,14 +67,14 @@ class HomeScreenActivity : AppCompatActivity() {
     lateinit var context: Context
     private val refreshDatabase by inject<RefreshDatabase>()
-    lateinit var appUpdateManager: AppUpdateManager
     lateinit var downloadManager : DownloadManager
```
DownloadManager is a system service which you could, but usually would not inject.

Ch4t4r commented on app/src/main/java/com/celzero/bravedns/ui/HomeScreenActivity.kt:
```diff
@@ -243,14 +236,23 @@ class HomeScreenActivity : AppCompatActivity() {
     }
 
+    fun checkForUpdate(userInitiation:Boolean = false) {
+        if (persistentState.downloadSource == Constants.DOWNLOAD_SOURCE_PLAY_STORE) {
```
Should this stay? With this PR the G-Play update check is only done for apps installed from G-Play, other builds (F-Droid in the future, from website) won't even contain the dependency for it and always use the web updater.
Currently I left this check in here for users who use the G-Play version but did not download from G-Play.

hussainmohd-a replied:
We don't need this shared pref value or check anymore.

ignoramous replied:
@Ch4t4r Hussain says, this check isn't required (we won't distribute PlayStore flavoured builds from anywhere else). But... PlayStore mirrors like ApkMirror and others might distribute that flavour... In which case, we would need to fallback on the web-updater, right, and not the PlayStore update checks? Or... How is this ideally supposed to work? 

Today, RethinkDNS distributed from the website and PlayStore are signed with the same keys.

Ch4t4r replied:
I had the mirrors in mind as well. You can't know whether the play services are present on those devices in advance -- you could test at runtime though and fallback to the web updater if they aren't.

ignoramous replied:
You're right. I guess we need to check for presence of Play Services too (which, if absent, would skip update-checks and show an error toast on Play flavoured builds): https://github.com/googlesamples/google-services/blob/7cbe0486697a/android/gcm/app/src/main/java/gcm/play/android/samples/com/gcmquickstart/MainActivity.java#L103-L117

Ch4t4r commented on app/src/play/java/com/celzero/bravedns/StoreAppUpdater.kt:
```diff
@@ -0,0 +1,96 @@
+                unregisterListener(listener)
+                Log.e(LOG_TAG, "no update or updating not allowed.")
+                if (isUserInitiated) {
+                    listener.onUpdateCheckFailed()
```
I copied this over from the old logic, but you might want to have another check here. Currently this'll tell the user the update check failed when he is up-to-date or in-app-updates aren't allowed currently (because the quota has already been reached)

ignoramous replied:
Can you please add that as a `TODO` in the code comment?

Ch4t4r replied:
It's only a few if-statements, I can add them. It was more of a "do you want to have it this way"-question

ignoramous replied:
> Currently this'll tell the user the update check failed when it is up-to-date or in-app-updates aren't allowed currently (because the quota has already been reached).

> It was more of a "do you want to have it this way"-question

In that case and if I understood you right, from what you said, it'd be right to show the right thing here that's less ambiguous and misleading, that is, either "the app is up-to-date" or "to try again later" (in case of quota overflow)?

Ch4t4r replied:
Done

Ch4t4r commented:
I couldn't find any other proprietary dependencies. I checked that it runs the web updater on `fdroid`-flavored builds (`./gradlew installFdroidDebug`) or `play`-flavored builds (`./gradlew installPlayDebug`) not actually installed from G-Play. On play-flavored builds installed from G-Play it calls the play service updater.

The web updater works as intended (checks in the background; shows dialog when up-to-date/check failed only when user clicked on `check for app updates`).
The gplay updater is called correctly but I could only verify the negative-case (no update availabe). You might want to test it using internal app sharing as described [here](https://stackoverflow.com/a/59266778/4400700).

Ch4t4r commented:
I'm not sure whether the prebuilt tun2socks is okay. Open-source dependencies on a private binary repository would not be okay as only a handful of repos are approved for F-Droid apps. In this case the library is included in the git repo however, which _might_ be fine? Anyone could verify the included binary in the git repo is compiled from the commit mentioned.

Ch4t4r commented:
If a prebuilt tun2socks would not be okay a more complex config file would have to be created in the F-Droid RFP repo which first builts the library and then copies it over, using the prebuilt directive (something like here https://gitlab.com/fdroid/fdroiddata/blob/master/metadata/chat.rocket.android.yml#L141)

ignoramous commented:
> I'm not sure whether the prebuilt tun2socks is okay. 

Apart from `outline-go-tun2socks`, isn't the guava lib is embedded as a binary (`jar`) as well? #198 

We should ideally build `outline-go-tun2socks` with RethinkDNS. May be `git submodules` is worth a shot: [github/working-with-submodules](https://github.blog/2016-02-01-working-with-submodules/)

Ch4t4r commented:
I totally missed the guava dep. Is there any reason why it is not pulled from maven?

ignoramous commented:
> I totally missed the guava dep. Is there any reason why it is not pulled from maven?

May be an artifact of Intra's early beginnings: https://github.com/Jigsaw-Code/Intra/commit/c20b993a4ace174e2a8aad38276ed1a1ca8079c4

May be it can simply be fetched from maven-central without issues and both those.`jar`s in the `app/lib` folder can be rid.

ignoramous commented on app/src/main/java/com/celzero/bravedns/service/AppUpdater.kt:
```diff
@@ -1,7 +1,3 @@
-import android.app.Activity
-
 /*
  * Copyright 2020 RethinkDNS and its authors
```
`2021` 🤟

Ch4t4r replied:
Done :P

ignoramous reviewed (approved):
To me, this PR is good to merge. Thanks for your time and effort, Daniel.

Note: Re: F-Droid, _at least_ two more changes required:

1. Fetch guava from maven central.
2. Move `outline-go-tun2socks` in as a `git submodule`.

Ch4t4r commented:
I removed the file dependency and added the maven one. App still compiles and country flags (that is what it is used for, right?) still show properly in query log.

Ch4t4r commented:
> 
>     2. Move `outline-go-tun2socks` in as a `git submodule`.
Would having a monorepo (subfolder/module for tun2socks) also be fine for you? That would be simpler and we could just depend on it using pure gradle (and wouldn't have to init/pull the submodule all the time)

hussainmohd-a commented:
Including guava along with firebase was having an known issue `Duplicate class com.google.common.util.concurrent exception`. That's the reason in Intra it was included as part of lib.
 [https://stackoverflow.com/questions/56639529/duplicate-class-com-google-common-util-concurrent-listenablefuture-found-in-modu](reference)

As we are not using any firebase dependency I think we can add as maven.

ignoramous commented:
> > ```
> > 2. Move `outline-go-tun2socks` in as a `git submodule`.
> > ```
> 
> Would having a monorepo (subfolder/module for tun2socks) also be fine for you? That would be simpler and we could just depend on it using pure gradle (and wouldn't have to init/pull the submodule all the time).
> I'm not sure whether Android projects can have non-Android modules though.

We are going to include WireGuard in this project at some point, too: Prefer that it comes in as a git-module. I am mostly basing it off of on how Orchid handles external but critical (security sensitive) cross langauge deps: https://github.com/OrchidTechnologies/orchid/blob/master/.gitmodules

I take, your concern is that gradle process won't be straight-forward? That is, Android Studio build integration would be convoluted more than necessary?

hussainmohd-a reviewed (approved):
Looks good for me.
